### ISSUE TITLE
Impossible inline fragments should raise a validation error?

### ISSUE THREAD

**Issue by syrusakbary** [graphql/graphql-js#485]
I've been working in optimizing the execution performance in `graphql-core` and as part of that I realized it was possible to do some weird combinations and still have a valid query.

Is this intentional?

```graphql
{
  node(id:"dmVoaWNsZXM6NA==") {
    id
    ...on Film {
      id
      ...on Node {
        id
        ...on Vehicle {
          id
        }
      }
    }
  }
}
```

[Query in SWAPI GraphiQL](http://graphql-swapi.parseapp.com/?query=%7B%0A%20%20node(id%3A%22dmVoaWNsZXM6NA%3D%3D%22)%20%7B%0A%20%20%20%20id%0A%20%20%20%20...on%20Film%20%7B%0A%20%20%20%20%20%20id%0A%20%20%20%20%20%20...on%20Node%20%7B%0A%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20...on%20Vehicle%20%7B%0A%20%20%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A)

**Comment by josephsavona** [comment]
I chatted with @leebyron about this recently since it came up in the context of generated Relay mutation queries. IIRC, the reason for this is to make working with fragments easier. If a fragment is defined:

```
fragment VehicleFragment on Node {
   id
   ... on Vehicle {
     make
  }
}
```

It should be usable anywhere that `Node` is a valid type. So 

```
{
  node(id:"dmVoaWNsZXM6NA==") {
    id
    ...on Film {
      id
      ...on Node {
        ...VehicleFragment
      }
    }
  }
}
```

Should be allowed - which is equivalent to your original example. Note that it could be quite a pain if this was considered invalid, as it would reduce fragment reusability. That said, it may be reasonable to validate inline fragments differently than regular fragments, though the inconsistency isn't ideal.

**Comment by josephsavona** [comment]
So I think there are two aspects to this:
- Query validation - the consistency and simplicity of the current rules seem overall well-balanced
- Optimization - separately from what is considered a well-typed/valid query, it's possible to write a query transformation that eliminates fragment spreads that can never match in practice, such as the inner `... on Vehicle`.

**Comment by leebyron** [comment]
I agree with you, @syrusakbary, that our validation algorithm is currently not strict enough.

In your example, what validation currently does is check "is it ok to use type Node in a context of type Film?" and then "is it okay to use type Vehicle in a context of type Node?" - both answers are yes. But a stricter version of this could instead see that these inline fragments are nested in which case the latter check would become "is it okay to use type Vehicle in a context of type Node AND type Film?" to which the answer would be no.

The reason we decided not to do this stricter evaluation right away is mostly for a thought that inline fragments and fragment definitions should behave as similarly as possible. 

To expand on your example:

```graphql
{
  node(id:"dmVoaWNsZXM6NA==") {
    id
    ...FilmFrag
  }
}

fragment FilmFrag on Film {
  id
  ...NodeFrag
}

fragment NodeFrag on Node {
  id
  ...VehicleFrag
}

fragment VehicleFrag on Vehicle {
  id
}
```

Here is the same query, but with all inline fragments taken out into fragment definitions. Should we apply the same suggested stricter validation rules? I would argue *no*. 

Why? Because we *can't know* how one of these fragments will be used in the future in a way that is legal, and it would not make sense to report a validation issue just because one of many possible uses is restricted in some way.
